lib/std/heap.zig

```diff
@@ -8,6 +8,141 @@ const c = std.c;
+    .thumb, .thumbeb, .arm, .armeb, .aarch64, .aarch64_be => 4 << 10,
+    // Explicitly only 4kb.
+    // https://refspecs.linuxbase.org/ELF/zSeries/lzsabi0_zSeries.html#AEN798
+    .s390x => 4 << 10,
+    // Source: Wikipedia "MIPS" and Linux mips/Kconfig.
+    .mips, .mipsel, .mips64, .mips64el, .loongarch32, .loongarch64 => 4 << 10,
+    // Source: csky/Kconfig only selects HAVE_PAGE_SIZE_4KB.
+    .csky => 4 << 10,
+    // Hexagon's Kconfig/page.h in Linux. Non-huge pages go past 256KB.
+    .hexagon => 4 << 10,
+    // Source: Zig's own libc page.h for arc.
+    .arc => 4 << 10,
+    // Source: Wikipedia "Page (computer memory)"
+    .powerpc, .powerpc64, .powerpc64le, .powerpcle => 4 << 10,
+    .riscv32, .riscv64 => 4 << 10,
+    .sparc => 4 << 10,
+    .sparc64 => 8 << 10,
+    // Rare architectures with little support.
+    else => @compileError("Does pageSize() apply to this architecture? If so, file an issue."),
+};
+
+var runtime_page_size = std.atomic.Value(usize).init(0);
+
+/// Runtime detected page size.
+pub inline fn pageSize() usize {
+    if (@inComptime()) {
+        @compileError("pageSize() must NOT be used in comptime. Use page_size variants instead.");
+    }
+    if (page_size == page_size_max) {
+        if (std.debug.runtime_safety) {
+            const size = queryPageSize();
+            assert(size == 0 or size == page_size);
+        }
+        return page_size;
+    }
+    const size = queryPageSize();
+    assert(size > 0);
+    return size;
+}
+
+// Runtime queried page size.
+fn queryPageSize() usize {
+    var size = runtime_page_size.load(.unordered);
+    if (size > 0) return size;
+    size = switch (builtin.os.tag) {
+        .linux => if (builtin.link_libc) @intCast(std.c.sysconf(std.c._SC.PAGESIZE)) else std.os.linux.getauxval(std.elf.AT_PAGESZ),
+        .ios, .macos, .watchos, .tvos, .visionos => blk: {
+            const task_port = std.c.mach_task_self();
+            // This may fail "if there are any resource failures or other errors".
+            if (task_port == std.c.TASK_NULL)
+                break :blk 0;
+            var info_count = std.c.TASK_VM_INFO_COUNT;
+            var vm_info: std.c.task_vm_info_data_t = undefined;
+            vm_info.page_size = 0;
+            _ = std.c.task_info(
+                task_port,
+                std.c.TASK_VM_INFO,
+                @as(std.c.task_info_t, @ptrCast(&vm_info)),
+                &info_count,
+            );
+            assert(vm_info.page_size != 0);
+            break :blk @as(usize, @intCast(vm_info.page_size));
+        },
+        .windows => blk: {
+            var info: std.os.windows.SYSTEM_INFO = undefined;
+            std.os.windows.kernel32.GetSystemInfo(&info);
+            break :blk info.dwPageSize;
+        },
+        else => if (builtin.link_libc and std.c._SC != void and std.c._SC.PAGE_SIZE != void and @TypeOf(std.c.sysconf) != void)
+            std.c.sysconf(std.c._SC.PAGE_SIZE)
+        else
+            0,
```

Comment:
Are we expecting to hit this path on any known platforms? Just wondering if a `@compileError()` would be better.

Reply:
Yes, it'll fail on a bunch, currently via an assertion when it cannot rely on the arch test. I can turn it into a compile error.